source/agora/consensus/Validation.d

```diff
@@ -956,6 +957,55 @@ public bool isValidEnrollment (const ref Enrollment enrollment,
     return isInvalidEnrollmentReason(enrollment, block_height, findUTXO) is null;
 }
 
+/*******************************************************************************
+
+    Check the validity of a pre-image information
+
+    Pre-image infomation is considered valid if:
+        - The keys for enrollment in two pieces of pre-image information are
+            same
+        - Height for a pre-image is greater than height for a current image
+        - A current image is same as n times hashed value of a pre-image
+            (n = difference of two heights).
+
+    Params:
+        new_image = The pre-image information to check
+        prev_image = The previous pre-image information
+
+    Returns:
+        `null` if the pre-image is valid, otherwise a string
+        explaining the reason it is invalid.
+
+*******************************************************************************/
+
+public string isInvalidPreimageReason (const ref PreimageInfo new_image,
+    const ref PreimageInfo prev_image, uint validator_cycle) nothrow @safe
+{
+    if (new_image.enroll_key != prev_image.enroll_key)
+        return "The pre-image has a wrong enrollment key";
+
+    if (new_image.height <= prev_image.height ||
+        new_image.height - prev_image.height >= validator_cycle)
+        return "the pre-image has a invalid height";
```

Comment:
II would actually split this into two returns. One with `<` and one with the check against `validator_cycle` so you can provide a more useful error message.